BOLT 7: Onion message support (features 38/39)

rustyrussell opened the pull request:
These use onion encoding for simple one-way messaging: there are no error returns.

Any reply is done with an enclosed blinded path, a-la @t-bast (in fact, this is based on #765)

Note that this defines the message system, not the contents of messages
(e.g. invoice requests from offers).

t-bast commented on 07-routing-gossip.md:
```diff
@@ -1105,6 +1106,47 @@ A->D's `update_add_htlc` message would be:
+1. type: 385 (`onion_message`) (`option_onion_messages`)
+2. data:
+    * [`1366*byte`:`onionmsg`]
+	* [`onion_message_tlvs`:`onion_message_tlvs`]
```
BTW seeing that makes me realize something about the TLV extension in `update_add_htlc`.
For blinded paths, the `update_add_htlc` will have a different size from "normal" `update_add_htlc`s.
That means their anonymity set will be reduced and network observers could unblind paths.

I don't have a solution yet, but we should try to make all `update_add_htlc`s the same size, otherwise it's somewhat useless to have a fixed-size onion in it...

rustyrussell replied:
You can pad with a no-reply ping for this?  Have to make sure the numbers are right though!

t-bast commented on 07-routing-gossip.md:
```diff
@@ -1105,6 +1106,47 @@ A->D's `update_add_htlc` message would be:
+
+## The `onion_message` Message
+
+1. type: 385 (`onion_message`) (`option_onion_messages`)
```
NB: you didn't add `option_onion_messages` in Bolt 9 yet, is that intentional?

rustyrussell replied:
Oops, will do.

t-bast commented on 04-onion-routing.md:
```diff
@@ -366,6 +367,84 @@ otherwise meets the amount criterion (eg. some other failure, or
+	- For the remaining `reply_path` `path`:
+	  - MUST set `node_id` to the blinded node id to encrypt the onion hop for.
+    - Within `reply_path` `path`:
+      - MUST set `enctlv` to the ChaCha20 encryption of a valid `encmsg_tlvs` containing exactly
```
ChaCha20-Poly1305?

rustyrussell replied:
Yes, with all-zero nonce.  I fixed to xref into the (not yet included!) blinded path spec.

jonatack commented on 04-onion-routing.md:
```diff
@@ -366,6 +367,86 @@ otherwise meets the amount criterion (eg. some other failure, or
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST include exactly one of `next_short_channel_id`, `next_node_id`
+    or `enctlv` indicating the next node.
+- For the the final node's `onionmsg_payload`:
```
s/the the/the/

dr-orlovsky commented:
Got really confused trying to understand how this proposal and https://github.com/lightningnetwork/lightning-rfc/pull/755 are related: is one a later version of the other?

rustyrussell commented:
> Got really confused trying to understand how this proposal and #755 are related: is one a later version of the other?

Oops, they were (almost!) identical. I closed the prior one since this one has comments.

I'm rebasing on master now, having applied the typo fixes (thanks!).

My only q is whether we should drop the `reply_path` and put that inside whatever the payload being delivered is.   For invoice requests we already have such a field, so it would be easy, and simplify the onion message proposal markedly.

The downside is that you can't reply with an error if the payload itself is unparsable...

rustyrussell commented:
Trivial squash and rebase on master.

rustyrussell commented:
Hmm, should we give this a real feature number?

rustyrussell commented:
Since we're now tentatively self-assigning incremental feature numbers, I'm switching this to 38/39.

kaloudis commented on 09-features.md:
```diff
@@ -40,7 +40,9 @@ The Context column decodes as follows:
 | 18/19 | `option_support_large_channel`   | Can create large channels                                 | IN       |                   | [BOLT #2](02-peer-protocol.md#the-open_channel-message) |
 | 20/21 | `option_anchor_outputs`          | Anchor outputs                                            | IN       | `option_static_remotekey` | [BOLT #3](03-transactions.md)         |
 | 22/23 | `option_anchors_zero_fee_htlc_tx` | Anchor commitment type with zero fee HTLC transactions   | IN       |                   | [BOLT #3][bolt03-htlc-tx], [lightning-dev][ml-sighash-single-harmful]|
+4<<<<<<< HEAD
```
you've got a meatball hanging here. it's creating a new row in the table

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,86 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+The writer:
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST include exactly one of `next_short_channel_id`, `next_node_id`
```
Why not require they be encrypted? It seems like a bunch of extra code to handle being optionally-encrypted when we could just require it. I think there should at least be a rationale section that describes why someone would want to send an unencrypted next-hop.

rustyrussell replied:
Agreed, this is now fixed.

thomash-acinq commented on 07-routing-gossip.md:
```diff
@@ -1119,7 +1120,59 @@ A->D's `update_add_htlc` message would be:
+- MUST populate the per-hop payloads as described in [BOLT 4](04-onion-routing.md#onion-messages).
+- SHOULD retry via a different route if it expects a response and
+  doesn't receive one after a reasonable period.
+- SHOULD set `len` to 1366 or 32834.
```
In BOLT 4, `hop_payloads` is said to be of length 1300. Shouldn't it be `len - 66` for onion messages?

TheBlueMatt replied:
I believe the length here *includes* the MAC, version byte, and ephemeral public key, so its 1300 + 66 in the constant overhead. Again, all of this would be 10x more clear if we just specified the full messages instead of trying to make references.

thomash-acinq replied:
I understand that there is a constant overhead of 66 bytes, what I'm saying is that in BOLT 4, it says that `hop_payloads` is **always** of length 1300 but I don't think it should be true in the case of onion messages. If the `len` here is equal to 32834, then the `hop_payloads` should be of length 32768, not 1300. That should be mentioned in BOLT 4 (I wanted to put my comment there but Github wouldn't let me comment on a line that was not modified).

TheBlueMatt replied:
Ah, then I think we're agreeing - we should just define the messages instead of making a reference to another (already defined) message and trying to indicate the changes to the other message.

thomash-acinq replied:
The `onionmsg` contains 66 bytes of version, public key and hmac. So if `len` is 32834, then the `hop_payloads` would be of length 32768. I guess that's where it comes from but as I see it it's just a big arbitrary number.

t-bast replied:
Of course, I'm dumb, I missed the 66 bytes.
As long as it's a `SHOULD`, not a `MUST`, I just wanted to understand the rationale, so it explains it thanks :+1:

TheBlueMatt commented on 07-routing-gossip.md:
```diff
@@ -1119,7 +1120,59 @@ A->D's `update_add_htlc` message would be:
+1. type: 387 (`onion_message`) (`option_onion_messages`)
+2. data:
+    * [`point`:`blinding`]
+    * [`u16`:`len`]
```
Note that this is duplicative with the bigsize for the onion length, though I'd much prefer to keep this and drop the bigsize, or maybe move this size down to there. There is zero reason we need to use a bigsize to save the one extra byte.

thomash-acinq replied:
From what I understood, the `onionmsg` here corresponds to the `onion_packet` type (04-onion-routing.md, line 153) that is always 1366 bytes for payment onion but may be of a different size for message onions. That's why we need to specify the size `len` here. The `onion_packet` type then contains a `hop_payload` of variable length and that's why we need the [`bigsize`:`length`] there (this `length` can't be computed from the previous `len` because of the presence of `filler` of unknown size).
The `onionmsg_payload` tlv stream only describes the content of this `hop_payload` for onion messages.

It took me many readings of the spec to reach this understanding (assuming it is correct), the relationship between the different types must be made a lot clearer.

TheBlueMatt replied:
Ah, right, its `hop_payload` vs `hop_payloads`. This is doubly confusing because the Onion Messages section only says its an onion "with an alternate `hop_payload`" without mentioning that, also the whole onion bytes thiggy is variable-length.

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,102 @@ otherwise meets the amount criterion (eg. some other failure, or
+### Onion Messages
+
+Onion messages have an onion with an alternate `hop_payload`
+format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
```
Please, no bigsize, just drop it entirely and use the 2 byte length in the `onion_message` struct.

rustyrussell replied:
This matches the modern onion format in htlc onions, which means no fresh parser to write.

TheBlueMatt replied:
Huh, its a fresh parser either way because a completely different TLV namespace? I guess your length read is at a different layer from the fields read, then?

rustyrussell replied:
Yes, we have wrap/unwrap routines, which just needed to be rejigged to allow a size param (rather than 1300 hardcoded).  Was trying to KISS...

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,102 @@ otherwise meets the amount criterion (eg. some other failure, or
 
+### Onion Messages
+
+Onion messages have an onion with an alternate `hop_payload`
```
Please drop the reference to `hop_payload` and just specify the whole message. It takes a ton of reading to maybe kinda understand how the message is laid out on the wire, instead of being able to look at something.

carlaKC replied:
Having recently scrolled around up and down to figure out the message format, strongly agree here.

rustyrussell replied:
Yeah, this gets simpler when we remove the legacy format altogether.  But I've rewritten it to duplicate the full description of the onion (with references back, since you don't need to reimplement it but should share code!).

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -1201,6 +1298,7 @@ The following is an in-depth trace of an example of error message creation:
 [fips198]: http://csrc.nist.gov/publications/fips/fips198-1/FIPS-198-1_final.pdf
 [sec2]: http://www.secg.org/sec2-v2.pdf
 [rfc8439]: https://tools.ietf.org/html/rfc8439
+[routeblinding]: https://github.com/lightningnetwork/lightning-rfc/blob/route-blinding/proposals/route-blinding.md
```
This needs to be in the repo if we want to reference it, presumably in this PR. I have a bunch of comments for this file to make clearer what's going on, but can't easily leave comments, and we shouldn't be referencing another branch of the same repo.

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,102 @@ otherwise meets the amount criterion (eg. some other failure, or
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST set `enctlv` to a valid `encmsg_tlvs` containing `next_node_id`.
+  - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
+    `ChaChaPoly-1305` encryption using an all-zero nonce).
```
Its really kinda awkward to use ChaCha+HMAC-SHA256 for the outer onion and ChaCha+Poly1305 for the inner one. I guess you decided that saving the 16 bytes per hop is that important? But if so, why not just also use ChaChaPoly for the outer onion as well?

rustyrussell replied:
Again, this is fixed already unless we want to rewrite all onion code.  The inner one is better chachapoly1305, because that we can control!

TheBlueMatt replied:
Heh, on our end almost everything is abstracted so that the decrypt is 5 lines of code, but I can still trivially replace the MAC call cause which method is called is not abstracted. I get that your code isn't that way, but man it makes for a really awkward spec. I guess there's no risk for the creator to do any collisions and saving 16 bytes is worth it, but it may be worth it on your end to see if you can tweak things a bit to make reuse 95% of the onion decrypt code without being boxed into ChaCha-SHA256.

t-bast replied:
As discussed during the last spec meeting, roasbeef provided a pretty good answer: crypto libraries usually don't let you use poly1305 without chacha20, which is why hmac-sha256 was initially chosen.

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,102 @@ otherwise meets the amount criterion (eg. some other failure, or
+2. data:
+    * [`point`:`node_id`]
+    * [`u16`:`enclen`]
+    * [`enclen*byte`:`enctlv`]
```
It is not specified at all what this is, nor does route blinding tell you what it is, as far as I can tell. The route blinding doc does not either, or at least if it does you don't specify where to look for it.

t-bast replied:
Actually now that I realize that onion messages uses route blinding on the way forward as well, it's not encrypted by the recipient but by the sender (but on the reply path it will be encrypted by the recipient), so my name sucks...

I cannot do better than `encrypted_data` then, wdyt?

TheBlueMatt reviewed (commented):
This needs to be fully specified. I left a few leading comments but, broadly, there seems to be about half of what is needed to actually write an implementation, and leaning on linking to the route blinding doc makes it more confusing :/

rustyrussell commented:
OK, I renumbered a field and added enctlv test vectors.  There should be more (particularly invalid ones!), and there should be full onion test vectors too.

thomash-acinq commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+    1. type: 12 (`next_blinding`)
+    2. data:
+        * [`point`:`blinding`]
+    1. type: 14 (`self_id`)
+    2. data:
+        * [`...*byte`:`data`]
```
Isn't the `self_id` intended for the final recipient? I thought that `enctlv` was only for intermediate nodes and `reply_path` only for the final recipient. If that's the case, then `self_id` should be in the `onionmsg_payload` next to `reply_path` and not nested inside a `enctlv` that is not needed for the final node.

rustyrussell replied:
enctlv is optional for the final recipient, but in practice you need it, to contain *some* identifying information (proposed here as `self_id`) as soon as you start using onion messages.  Otherwise you need to remember the blinding factor for every single path you've ever sent out, to avoid responding inappropriately.

thomash-acinq replied:
I understood why you want the `self_id`, I just thought it would be simpler to put it directly in the `onionmsg_payload`, not in the `enctlv`. I've now realized that it needs to be in the `enctlv` from the `reply_path` so that the `self_id` can't be modified by the person who replies.

thomash-acinq commented on 01-messaging.md:
```diff
@@ -49,7 +49,7 @@ The messages are grouped logically into five groups, ordered by the most signifi
   - Channel (types `32`-`127`): messages used to setup and tear down micropayment channels (described in [BOLT #2](02-peer-protocol.md))
   - Commitment (types `128`-`255`): messages related to updating the current commitment transaction, which includes adding, revoking, and settling HTLCs as well as updating fees and exchanging signatures (described in [BOLT #2](02-peer-protocol.md))
-  - Routing (types `256`-`511`): messages containing node and channel announcements, as well as any active route exploration (described in [BOLT #7](07-routing-gossip.md))
+  - Routing (types `256`-`511`): messages containing node and channel announcements, as well as any active route exploration or messaging (described in [BOLT #7](07-routing-gossip.md))
```
What is the rationale for putting messages in the routing category?
I don't think it belongs to any of the current categories and that's fine, it should be on its own.

rustyrussell replied:
We could renumber it to 513, I guess.

thomash-acinq commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+    - if this is a reply to an onion message which was sent with a `reply_path`:
+      - MUST respond (or not respond) exactly as if it did not send the initial onion message.
+  - if it wants to send a reply:
+    - MUST create an onion encoding using `reply_path`.
```
This needs to be explained with more details.

rustyrussell replied:
Yes...

thomash-acinq commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+    1. type: 12 (`next_blinding`)
+    2. data:
+        * [`point`:`blinding`]
+    1. type: 14 (`self_id`)
```
I don't think `self_id` is an informative name. I think it should be `path_id` because it identifies the reply path.

rustyrussell replied:
Sure.

thomash-acinq commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+    - MUST ignore `padding`, if any.
+    - SHOULD forward the message using `onion_message` to the next peer indicated by `next_node_id`.
+    - if it forwards the message:
+      - if `blinding` is specified in the `enctlv`:
+         - MUST pass that as `blinding` in the `onion_message`
```
Why not always use the derived `blinding`? What is the use-case where we need to override the default `blinding`?

rustyrussell replied:
If I give you a blinded path from Alice, you need to make an onion which goes to Alice then uses the blinded path.  This requires blinding override.

t-bast replied:
Can you detail that @rustyrussell? I don't understand it either.
If for example Carol creates a blinded path `Alice -> Bob -> Carol`.
Carol provides a list of `encrypted_recipient_data` and a first `blinding` point that must be sent to Alice.
But that `blinding` cannot be inside the `encrypted_recipient_data` as it's required to actually decrypt it...?
It seems to me that you just need the `blinding` tlv fields that is part of the `onion_message` for that.

thomash-acinq replied:
I receive the following blinded route built by Carol: `(blinding0, Alice, (BlindedAlice, BlindedAliceData) -> (BlindedBob, BlindedBobData) -> (BlindedCarol, BlindedCarolData))` and I want use it by adding Dave in front.

I want to create a route `(blinding1, Dave, (BlindedDave, BlindedDaveData) -> (BlindedAlice, BlindedAliceData) -> (BlindedBob, BlindedBobData) -> (BlindedCarol, BlindedCarolData))`. For that I generate randomly `blinding1` and use it to blind `(Dave, DaveData)`, I then obtain `(BlindedDave, BlindedDaveData)` and a derived `blinding1Next`. However `blinding1Next` is not equal to `blinding0`, so I need to specify `blinding0` as an override for the rest of the route. I pass this `blinding0` to Dave, inside `DaveData` and Dave MUST pass this `blinding0` as `blinding` in the `onion_message` it sends to Alice.

t-bast replied:
If you're concatenating two blinded routes I can see why you would need that, but is there a use-case for that though?
In all scenarios I can think of you would use a normal route through Dave, the blinded part would only start at Alice.
The recipient is the one that wants to hide, why would the sender would use a blinded route for the part that he chooses?

thomash-acinq replied:
There is no such thing as a normal route. Everything is blinded.

t-bast replied:
Right, that makes sense! It's worth explaining a bit more in the PR then, the rationale section should make this obvious, since route blinding is designed to prepend a normal route to a blinded one (at least for payment scenarios).

thomash-acinq commented on bolt04/enctlvs.json:
```diff
@@ -0,0 +1,61 @@
+    "rho": "d6297b5645c27be8ea1240cc96e162c53fd9c389c6bec25cbdb4b6b3f654e7bc",
+    "enctlv_hex": "8285acbceb37dfb38b877a888900539be656233cd74a55c55344fb068f9d8da365340d21db96fb41b76123207daeafdfb1f571e3fea07a22e10da35f03109a0380b3c69fcbed9c698086671809658761cf65ecbc3c07a2e5"
+  }
+]
```
It would be useful to have the whole onion combining these 3 payloads.

rustyrussell replied:
Yes, agreed.

t-bast replied:
I agree, I think most of these test vectors are just redundant with the test vectors in https://github.com/lightning/bolts/pull/765, onion messages should instead add test vectors for the onion-message-specific bits (a complete `onion_message` with an onion in it, and keys to decrypt it at intermediate nodes and at the recipient).

rustyrussell replied:
Yes, with those test vectors I think this should be reduced to the same form (ideally, same data) as those with onionmsg specifics.  I'll do that tomorrow...

TheBlueMatt commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+An intermediate node expects an `enctlv` which it can decrypt using
+the `blinding` which it is handed along with the onion message.
+
+1. `tlv_stream`: `onionmsg_payload`
```
I think I mentioned this off GitHub, but I really think there needs to be a separate TLV stream for the actual message content. Shoving it in the same one is cool, but in practice if we have multiple message types we're gonna want to pass a whole TLV stream to different parsers, not have each message type hook the same TLV stream parsing logic.

rustyrussell replied:
You mean we have a `payload` blob of bytes at this level, and YA tlv inside it?  Possible, but I don't think demux is going to get any prettier?

TheBlueMatt replied:
That's one way to accomplish it, yes. I don't really understand how to demux without some second-level thing here - we have a tlv stream that contains fields that the onion message decoder/processor needs to handle, and fields that the onion-message-type handler needs to process. Presumably those need to be two separate things because there could be different types of onion messages. Arguably there needs to be, per type of onion message, a different, single TLV record in the payload here that declares the type and has all the type-specific data (eg type 1 would be "BOLT 12 offer request" and type 3 would be "BOLT 12 offer response", and....etc).

rustyrussell replied:
Perhaps we just say "64 and above"?  It is much nice to use one number per message type, IMHO, rather than one per feature.

t-bast replied:
SGTM, let's keep it like this for now and we can have this discussion when we review features that actually add tlvs for the final node.

rustyrussell commented on bolt04/enctlvs.json:
```diff
@@ -0,0 +1,61 @@
+    "blinding": "03fc5e56da97b462744c9a6b0ba9d5b3ffbfb1a08367af9cc6ea5ae03c79a78eec",
+    "encmsg": {
+      "next_node_id": "027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa2007",
+      "blinding": "070707070707070707070707070707070707070707070707070707070707070701"
```
I think 02989c0b76cb563971fdc9bef31ec06c3560f3249d6ee9e5d83c57625596e05f6f is the pubkey of 0707070707070707070707070707070707070707070707070707070707070707 ?  The reader gets the pubkey, the writer needs to privkey...

t-bast reviewed (commented):
Sorry for the late review!
I think it's mostly good, some details need to be clarified but I like this proposal.
It's also quite a small spec addition, and it's well contained into specific sections.

rustyrussell commented:
OK, I squashed all the previous fixes into a single commit, and rebased on #765!

I have also renumbered it to 513 from the old number in the gossip range, and moved the message definition into BOLT 4.

You're right the old test vectors are redundant, but I'll simply update them for now, once I've updated c-lightning!

joostjager commented on 09-features.md:
```diff
@@ -41,6 +41,7 @@ The Context column decodes as follows:
 | 20/21 | `option_anchor_outputs`          | Anchor outputs                                            | IN       | `option_static_remotekey` | [BOLT #3](03-transactions.md)         |
 | 22/23 | `option_anchors_zero_fee_htlc_tx` | Anchor commitment type with zero fee HTLC transactions   | IN       |                   | [BOLT #3][bolt03-htlc-tx], [lightning-dev][ml-sighash-single-harmful]|
 | 26/27 | `option_shutdown_anysegwit`         | Future segwit versions allowed in `shutdown`              | IN       |                   | [BOLT #2][bolt02-shutdown]   |
+| 38/39 | `option_onion_messages`        | Can forward onion messages                                | IN9        |                   | [BOLT #7](07-routing-gossip.md#onion-messages) |
```
Why is the forwarding capability defined as an invoice feature?

t-bast replied:
I guess it's the description that's incorrect, it should be a more general "Supports onion messages"? You need to put that into the invoice so that senders know they can send you onion messages if they don't have your `node_announcement`.

joostjager replied:
If you accept onion messages, but don't forward them, how is this communicated? Only flag the bit in the invoice and not in the node announcement?

t-bast replied:
> Only flag the bit in the invoice and not in the node announcement?

Yes that could indeed be a good way of communicating this!
Do you think it's insufficient?

joostjager replied:
Actually, is there even a reason for advertising the capability in the invoice at all? For bolt12, the messaging is already completed when you've got the invoice.

rustyrussell replied:
Well, if you give someone a bolt12, you're kind of implying you can talk to them.  I agree with Joost, it's a weird thing to put in an invoice at all.

t-bast commented on 04-onion-routing.md:
```diff
@@ -525,6 +527,93 @@ same length. It's particularly useful when adding dummy hops at the end of a
+        * [`point`:`first_node_id`]
+        * [`point`:`blinding`]
+        * [`...*onionmsg_path`:`path`]
+    1. type: 10 (`encrypted_data_tlv`)
```
We're currently using 4 in eclair:

```suggestion
    1. type: 4 (`encrypted_data_tlv`)
```

I don't see a good reason to leave a gap here, it's a bit confusing because it looks like the tlv namespace is shared between `onionmsg_payload` and `encrypted_data`, which isn't the case.

rustyrussell replied:
Yep, I came here to change that :)

rustyrussell commented:
OK, minor tweaks.  I've not put the meta-suggestion on how to assign payload numbers, but I've made it clear it's 64+ (we may eventually assign 1024+ for experimental I guess?).

- onion_msgpayload encrypted_data_tlv is type 4, not 10.
- enctlv in path is now called `encrypted_recipient_data` which is consistent.
- Removed setting of onionmsg feature in invoices.

vincenzopalazzo reviewed (approved):
LGTM https://github.com/lightning/bolts/pull/759/commits/6e99c5feaf60cb797507d181fe583224309318e9

t-bast reviewed (commented):
Looks mostly good to me.

I created a commit with a few fixes / clean-up here: https://github.com/t-bast/bolts/commit/5ba24ce39f93c84fe136c6c0d875418db694874c
Feel free to take only the parts you agree with if necessary.

The only remaining things to address from my point of view are the test vectors.
The current test vectors you added are only testing route blinding, not onion messages, so I feel they should be removed in favor of the existing route blinding test vectors (which we can improve / change as per discussions in #765).

IMO the test vector we include here should be a complete `onion_message` over a few hops, with a `reply_path` and a matching response `onion_message` from the recipient. I'll sync with @thomash-acinq on that, maybe we can propose such a test vector ourselves.

thomash-acinq commented:
> IMO the test vector we include here should be a complete `onion_message` over a few hops, with a `reply_path` and a matching response `onion_message` from the recipient. I'll sync with @thomash-acinq on that, maybe we can propose such a test vector ourselves.

@rustyrussell has already sent me such test vectors (you can see them here: https://github.com/ACINQ/eclair/pull/2065) but I think he wanted to rewrite these test vectors to use the exact same scenario as in the blinded routes test vectors.

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+    - `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = SHA256(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
```
Since we made the ECDH explicit in line 430 we could also here help a bit by writing:

```
E(i+1) = e(i+1)*G 
       = SHA256(E(i) || ss(i)) * e(i) * G
       = SHA256(E(i) || ss(i)) * E(i)
```

with the comment from line 427 everyone can see that the math is correctly

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+#### Requirements
+
+A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:
```
Just to be clear: `N(0)` is the "introduction node"? Maybe we could add this similar to the other explanations in the following lines.

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+- If it creates `encrypted_data`:
+  - MUST encrypt it with ChaCha20-Poly1305 using the `rho(i)` key and an all-zero nonce
+  - MAY store data in the `path_id` of the payload to itself to verify that
+    the route is used in the right context
```
How does that verification work? What is meant with "right context"? I see that there was already rational explaining this so maybe here it is a bit redundant?

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+  - MAY store data in the `path_id` of the payload to itself to verify that
+    the route is used in the right context
+  - MAY add padding data to ensure all `encrypted_data` have the same length
+- MUST communicate the blinded node IDs `B(i)` and `encrypted_data(i)` to the sender
```
Why?
Given `N(i)` which you require the node can also compute the `B(i)` with the help of `E(i)`

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of
```
See comments below for more context but I don't understand the extra use coming from the additional complexity of the blinded node id. Is that just a name for the variable to explain the formulas?

The actual encrypted node_ids need to be stored in `next_node_id` anyway and the blinded id can always be computed from the ephemeral shared secret and the actual node id (as currently specified in line 473 to 475) but not the other way around. So the blinded node ids seem to be useless to me.

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+- MUST compute:
+  - `ss(i) = SHA256(k(i) * E(i))` (standard ECDH)
+  - `b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)`
```
If `B(i)` was actually explicitly communicated to `N(i)` the node could verify that `B(i) = b(i)*G` but other than that I do not see any use of the `B(i)`

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = SHA256(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
+- SHOULD create an `encrypted_data` for each blinded node `B(i)`
```
Just to be clear `next_node_id` should be `N(i+1)` and not `B(i+1)`. It is strange to me to call the nodes on the blinded path `N(0)--> N(r)` now `B(i)`.

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+last node of the first route needs to know the first `blinding` of the second
+route: the `next_blinding_override` field must be used to transmit this data.
+
+The final recipient must verify that the blinded route is used in the right
```
why is "must" not written capitalized here?

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(1)` as the next blinding point
```
While `E(i)` makes a lot of sense the notation blinding point is actually a bit confusing as I initially thought `B(1)` which as the blinded node_id is also a point. But I guess this all boils down to the general gist / suggestions to remove the notation of a blinded node_id `B(i)`

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+When concatenating two blinded routes generated by different nodes, the
+last node of the first route needs to know the first `blinding` of the second
+route: the `next_blinding_override` field must be used to transmit this data.
```
Given I understand this correctly I think this is a neat trick to do the ephemeral key switch!

lightning-developer commented on 04-onion-routing.md:
```diff
@@ -1143,9 +1390,59 @@ The _origin node_:
+
+The reader:
+- MUST handle the per-hop payloads as described in [Onion Message Payload Format](onion-message-payload-format).
+- SHOULD accept onion messages from peers without an established channel.
```
I assume this is with Offers in mind but it is strange to me. Given the ongoing discussion about spam prevention and rate-limiting it seems reasonable to require at least a payment channel before accepting and forwarding onion messages

rustyrussell replied:
You don't have to forward them, or you can ratelimit them, but you should definitely accept them.  Like you do pings...

rustyrussell replied:
Noted in rationale.

lightning-developer reviewed (changes requested):
I went through the ECDH part of the route blinding and it looks good and correct to me. I am only a bit confused about notation of a blinded node id (see comments). The blinded node_id `B(i)` is confusing for me and seems unnecessary as the `N(i)` still needs to be known and the `E(i)` is included in the onions from which the `B(i)` can be computed anyway.

I also added a few stylistic suggestions to make it easier for future readers (feel free to ignore them)

Additionally given the context of rate-limiting and spam prevention I was wondering if we should not require an existing payment channel to send / receive an onion message from a peer?

lightning-developer commented:
May I also add to the spam prevention discussion that there is the article: "Point Break: A Study of Bandwidth Denial-of-Service Attacks against Tor" ([paper](https://www.usenix.org/system/files/sec19-jansen.pdf), [video presentation](https://youtu.be/pSeIy4y0o3c
)) which suggests that bandwidth denial of service attacks work if the attacker requests large files but does not read on the TCP-socket to the exit node. in this way the attacker can - for relatively low bandwidth cost - produce a large load on the network. I believe that the small fixed size for onion messages doesn't yield a similar risk for cheap bandwidth denial of service attacks as the attacker would have to pay roughly for the same bandwidth. 

For additional context I wish to add that TOR has - as you all probably know - an entire DoS Mitigation Subsystem which is described at 
 https://gitlab.torproject.org/legacy/trac/-/issues/24902 and https://www.mail-archive.com/tor-dev@lists.torproject.org/msg11628.html

t-bast commented:
@lightning-developer thanks a lot for the route blinding review!

If it's not too annoying, could you please move the comments that are only related to route blinding to its dedicated PR (#765) so that we keep discussions here about onion messages only?

I'll gladly answer your route blinding feedback on #765

joostjager commented:
>Additionally given the context of rate-limiting and spam prevention I was wondering if we should not require an existing payment channel to send / receive an onion message from a peer?

If you do this, you can't deliver an onion message directly to a destination if you don't have a channel with them. You'd need to rely on forwarding nodes being available.

lightning-developer commented:
> > Additionally given the context of rate-limiting and spam prevention I was wondering if we should not require an existing payment channel to send / receive an onion message from a peer?
> 
> If you do this, you can't deliver an onion message directly to a destination if you don't have a channel with them. You'd need to rely on forwarding nodes being available.

Oh I see the misunderstanding. I thought this was meant for forwarding and processing of onion messages. Let's say we have a channel graph `A-->B-->C-->D` and `A` wanted to send an onion message to `D` I thought this would be possible. In particular `D` SHOULD accept the message even though it is not a direct peer of `A`. If we however had another node `M` peering with `A` with no channel I thought `M` should not be able to send an onion message via `A` to other nodes.

instagibbs commented on 04-onion-routing.md:
```diff
@@ -1143,9 +1390,59 @@ The _origin node_:
+2. data:
+    * [`point`:`blinding`]
+    * [`u16`:`len`]
+    * [`len*byte`:`onionmsg`]
```
can we call this `onionmsg_payloads` and refer to it in the section mentioning `hop_payload` to make the linkage explicit?

rustyrussell replied:
I called them both "onion_message_packet" to be clear...

valentinewallace commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
```
As discussed on the BOLT 12 Telegram, it may make sense to get rid of this field since `next_node_id` is currently mandatory, thus making the scid redundant. In addition, IIUC Acinq will currently error if the `next_node_id` field is not present (cc @t-bast).

t-bast replied:
I don't think we should remove it. For payments we still want to use the `short_channel_id` to identify the next node as it uses less space in the onion. For onion messages we're only using `next_node_id`, but for blinded payments we use `short_channel_id`.

t-bast replied:
We can calculate that, but I'd rather just keep this option of using less space, it can always be useful (e.g. to add more dummy hops at the end of a route). This is what we use for normal payments relay, so it shouldn't require implementations to add any new code, so I don't see why we'd want to remove it?

valentinewallace replied:
> I don't think we should remove it. For payments we still want to use the `short_channel_id` to identify the next node as it uses less space in the onion. For onion messages we're only using `next_node_id`, but for blinded payments we use `short_channel_id`.

That seems fair. Based on that, I'd prefer if we made `next_node_id` explicitly mandatory for onion messages and disallow `scid`, and the reverse requirements for payment relay. (I believe that's what eclair does now, at least the `next_node_id` part?)

valentinewallace commented on bolt04/enctlvs.json:
```diff
@@ -0,0 +1,61 @@
+    "test name": "Simple enctlv for Alice, next is Bob",
+    "node_privkey": "414141414141414141414141414141414141414141414141414141414141414101",
+    "node_id": "02eec7245d6b7d2ccb30380bfbe2a3648cd7a942653f5aa340edcea1f283686619",
+    "blinding_secret": "050505050505050505050505050505050505050505050505050505050505050501",
```
@rustyrussell I'm wondering why this is 33 bytes? Aren't secrets always 32 bytes?

rustyrussell replied:
It's the bitcoin-style notation for private keys.... append 01.  Horrible, let's kill it...

valentinewallace commented on bolt04/enctlvs.json:
```diff
@@ -0,0 +1,61 @@
+    "blinding": "03fc5e56da97b462744c9a6b0ba9d5b3ffbfb1a08367af9cc6ea5ae03c79a78eec",
+    "encmsg": {
+      "next_node_id": "027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa2007",
+      "blinding": "070707070707070707070707070707070707070707070707070707070707070701"
```
Could we rename this `next_blinding_override` for clarity?

thomash-acinq replied:
I agree with `next_blinding_override`. `next_blinding_override_secret` is too much though, we don't want to have `next_node_id_secret` either.

valentinewallace replied:
Okay, if we called it `next_blinding_override` I would expect it to be a point, not a secret, bit confusing. Maybe we can have an inline comment like `// secret` after the json field. Not a huge deal though.

rustyrussell replied:
I'd prefer more verbosity than less, so I think I side with @valentinewallace here...

valentinewallace commented on bolt04/enctlvs.json:
```diff
@@ -0,0 +1,61 @@
+    "E": "03fc5e56da97b462744c9a6b0ba9d5b3ffbfb1a08367af9cc6ea5ae03c79a78eec",
+    "H(E || ss)": "9afb8b2ebc174dcf9e270be24771da7796542398d29d4ff6a4e7b6b4b9205cfe",
+    "next_e": "5149cb0ce6100209f9a2a1140c4e6e8235381aac6201da2c0f8ce1f91bcc3b6e01",
+    "encmsg_hex": "0421027f31ebc5462c1fdce1b737ecff52d37d75dea43ce11c74d25aa297165faa20070c2102989c0b76cb563971fdc9bef31ec06c3560f3249d6ee9e5d83c57625596e05f6f",
```
Believe this encodes `next_blinding_override` as TLV type 12, but the TLV type in the spec is 8

GeorgeTsagk commented:
AFAIU;
The proposed onion messages use an onion routing mechanism similar to the one payments use to forward HTLCs.
But the flow of forwarding an onion message is not related somehow to any amount forwarding. Why would a node opt-in to this? Other nodes might eat up your bandwidth and you don't have something to earn from that.

t-bast commented:
> But the flow of forwarding an onion message is not related somehow to any amount forwarding. Why would a node opt-in to this? Other nodes might eat up your bandwidth and you don't have something to earn from that.

For many reasons:

- it's extremely cheap (no storage required, low cpu costs and messages below 65kb)
- it's trivial to rate-limit (we will publish a simple and efficient scheme soon)
- it helps your node attract more payments, which will earn you fees (e.g. by helping users fetch bolt 12 invoices that they will then pay)
- it's ok if some nodes are unhappy with these trade-offs and don't turn the feature on

GeorgeTsagk commented:
> For many reasons:
> 
> * it's extremely cheap (no storage required, low cpu costs and messages below 65kb)
> * it's trivial to rate-limit (we will publish a simple and efficient scheme soon)
> * it helps your node attract more payments, which will earn you fees (e.g. by helping users fetch bolt 12 invoices that they will then pay)
> * it's ok if some nodes are unhappy with these trade-offs and don't turn the feature on

All four arguments are very strong. I want to focus a bit more on the 3rd one as far as rewards go. Can you elaborate more (or possibly link to a document) explaining this payment attraction? Does the route carrying the bolt12 invoice also bind a route for the payment?

But apart from that, if one wants to use the onion messaging to transmit data related to a use-case of their own (unrelated to any LN operation) then there is no direct compensation to the nodes forwarding that traffic. Yes this can be rate limited, but that's more of a protection measure from the side of the routing node, exposing the node sending the messages to unreliable forwarding. I also think this PR is not introducing some kind of receipt to let the message sender know if a message arrived or not, which can also help make things complicated from the sender's PoV.

So I suppose in a next iteration of onion messaging, perhaps in an attempt to make it more reliable, a reward system would have to be introduced. This would incentivize nodes not only to opt-in to this feature, but also provide very good services by i.e. maintaining a very stable/fast internet connection.

Mostly I want to wrap this up by comparing two possible paths one can follow to send (generic purpose) messages (of relatively small size) over Lightning:

- Onion messaging; requires protocol update, lightweight forwarding, no forwarding reward, requires whole route to opt-in, (?) no receipt for message arrival, (?) would require introduction of a new reward system for reliability
- Messaging over payments: no protocol update, dirtyish forwarding (DBs can optimize), forwarding reward is payment reward (effectively the base fees), no need for route opt-in, preimage is receipt, costs

In case second bullet wasn't clear, I'm referring to attaching custom TLV records to the final destination of a payment. One of the reasons I'm focusing so much on direct reward is because I truly believe that this will scale the adoption and make messaging more reliable, regardless of which path one chooses.

renepickhardt commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+1. `tlv_stream`: `onionmsg_payload`
+2. types:
+    1. type: 2 (`reply_path`)
```
How is this secure from spoofing the `reply_path`? 

As the originator from the Onion Message is not known couldn't one basically create a `reply_path` to any node in the network and create potential DoS attacks (especially if rate limiting of onion messages is being introduced)

thomash-acinq replied:
You would have to send one message to a random node so that your target is spammed with one message. How is that better than just sending the message to your target directly? You can indeed do that but it doesn't amplify the number of messages.

For a DoS attack you would more likely want to have as many hops as possible to amplify the attack but this reflection doesn't give you anything interesting.

renepickhardt replied:
True, unless I find a way in which the node will send the reply via the `reply_path` several times because it thinks the reply has not been received as we use an unreliable protocol (this could eventually happen from application land)

renepickhardt replied:
I suppose this will work. Though it seems still a bit risky

renepickhardt reviewed (commented):
Given [this conversation](https://twitter.com/renepickhardt/status/1534552974311161856) with @fiatjaf made me realize that we **might** have a DoS attack vector being introduced with the blinded `reply_paths` in onion messages (at least if my current mental model of how they work is accurate) .

## The attack Scenario

Let's say I wish to DoS the node of @rustyrussell I would send an onion message to every node of the network and attach a blinded `reply_path` that goes to Rusty's node. This would either put a heavy load on Rusty's node or it would put heavy load to his inbound channels (which if we included rate limiting to onion messages would probably congest them). Of course this assumes that nodes will just answer incoming onion messages to the given reply path (which I guess in the case of offers is very application specific)

## Comparison with TCP/IP 
In the TCP/IP world [IP Spoofing](https://en.wikipedia.org/wiki/IP_address_spoofing) is mitigated by packet filtering and [ingress filtering](https://en.wikipedia.org/wiki/Ingress_filtering) but I think this is not possible as we neither have a logical network and address space nor the ability to look inside the packets due to our onion routing construction and blinding.

## Mitigation ideas

The best way I can currently think of to mitigate this is to use the onion errors that we have in onion routing for the reply meaning that onion messages would not be fire and forget but putting load on routing nodes to be able to send back errors by remembering where they came from. I guess as no HTLCs are being attached to onion messages one could have a time to live of 2 seconds per each hop on the path for an onion message (and potentially add a few seconds to disallow for timing attacks) which would probably limit the data storage requirement of routing nodes to about 30 seconds.

That being said I don't like the mitigation as I believe it introduces many challenges.

carlaKC commented on 04-onion-routing.md:
```diff
@@ -525,6 +527,93 @@ same length. It's particularly useful when adding dummy hops at the end of a
+
+The writer:
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST set `enctlv` to a valid `encrypted_data_tlv` stream containing `next_node_id`.
```
`enctlv` has been renamed in the format above but not in the requirements section

rustyrussell replied:
Oops, fixed!

valentinewallace commented on 04-onion-routing.md:
```diff
@@ -525,6 +527,93 @@ same length. It's particularly useful when adding dummy hops at the end of a
+        * [`point`:`first_node_id`]
+        * [`point`:`blinding`]
+        * [`...*onionmsg_path`:`path`]
+    1. type: 4 (`encrypted_data_tlv`)
```
I like `encrypted_tlv_stream` or `encrypted_tlvs` but perhaps overly explicit lol

rustyrussell replied:
The route blinding calls this `encrypted_data`, which gets decrypted into a `encrypted_data_tlv`.  Lets' be consistent: if we don't like it we can fix all of it together...

valentinewallace commented on 04-onion-routing.md:
```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+2. types:
+    1. type: 2 (`reply_path`)
+    2. data:
+        * [`point`:`first_node_id`]
```
Discussed a while ago on [telegram](https://t.me/bolt12org/1687) -- do we want to get rid of this field and put the unblinded `first_node_id` in the blinded path's first hop? I believe offers encoding does this currently

rustyrussell replied:
I thought @thomash-acinq pointed out that you can't do this without knowing the blinding secret :(  I think that may be correct...

valentinewallace replied:
Ah! That checks out to me, thanks

rustyrussell commented on 04-onion-routing.md:
```diff
@@ -367,6 +368,152 @@ otherwise meets the amount criterion (eg. some other failure, or
+  - if the `enctlv` contains `path_id`:
+    - MUST drop the message.
+  - otherwise:
+    - MUST ignore `padding`, if any.
```
You'd *think* so, but Id' rather say *exactly* what to do with each field, so...

rustyrussell commented:
OK, rebased on latest https://github.com/lightning/bolts/pull/765, folded all previous fixes, and added a commit which addresses all previous feedback.

I need to rework the test vectors: route blinded has excellent vectors and so we just need incremental ones on that.

rustyrussell commented:
Rebased on the latest #765, and added three new commits:
1. Rework it into a single section, rather that splitting the "onion payload" from the "onion message" itself.
2. Add a quick note that we don't use associated data (unlike payments, which use the payment_hash)
3. Remove old test vectors, insert shiny new *complete* ones in same style as blinded payment onion tests.

rustyrussell commented:
Rebased again!  And added a final commit:
1. Changed the type of reply_path to use the bolt12 type (thanks @jkczyz).
2. Renamed onionmsg_path to more accurate onionmsg_hop.

rustyrussell commented:
Fairly trivial rebase on top of updated https://github.com/lightning/bolts/pull/765, and fixup to make test vector reproducible!

dunxen commented on 04-onion-routing.md:
```diff
@@ -1131,6 +1395,177 @@ The _origin node_:
+onion messages.
+
+`len` allows larger messages to be sent than the standard 1300 bytes
+allowed for an HTLC onion, but this should be used sparingly as it is
+reduces anonymity set, hence the recommendation that it either look
```
nit:
```suggestion
allowed for an HTLC onion, but this should be used sparingly as it
reduces the anonymity set, hence the recommendation that it either looks
```

rustyrussell replied:
Applied, thanks!

valentinewallace commented on bolt04/blinded-onion-message-onion-test.json:
```diff
@@ -0,0 +1,143 @@
+        "comment": "Bob->Carol",
+        "blinding_secret": "0101010101010101010101010101010101010101010101010101010101010101",
+        "tlvs": {
+          "padding": "0000000000000000000000000000000000000000000000000000",
```
Is there a reason to have padding in every hop of Dave's blinded path? It seems like the largest one wouldn't need padding, only the smaller ones, to make them all equal-length

rustyrussell replied:
Yes, that would have been clever.  Since I'm rebasing anyway, let me take a look...

rustyrussell replied:
Ah, we did it so that Alice wouldn't need to pad, but Dave wouldn't know that.  Fixing...

rustyrussell commented on 04-onion-routing.md:
```diff
@@ -1131,6 +1395,177 @@ The _origin node_:
+`encrypted_tlv_stream` which it can decrypt into an `encrypted_data_tlv`
+using the `blinding` which it is handed along with the onion message.
+
+Field numbers 64 and above are reserved for payloads for the final
+hop.
```
No, there is no such *requirement*, and I don't think we need to box ourselves in on that.  Just a convention.

I'll add a note.

rustyrussell commented:
Changes applied, and rebased.

rustyrussell commented:
OK, I squashed into a single commit and fixed some whitespace (and changed a "MUST send an error" to "MUST ignore" for consistency).

This now applies directly onto master now blinded payments are merged!

t-bast commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,178 @@ The _origin node_:
+- if `encrypted_data_tlv` contains `allowed_features`:
+  - MUST ignore the message if:
+    - `encrypted_data_tlv.allowed_features.features` contains an unknown feature bit (even if it is odd).
+    - the payment uses a feature not included in `encrypted_data_tlv.allowed_features.features`.
```
```suggestion
    - the message uses a feature not included in `encrypted_data_tlv.allowed_features.features`.
```

t-bast commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,178 @@ The _origin node_:
+onion messages.
+
+`len` allows larger messages to be sent than the standard 1300 bytes
+allowed for an HTLC onion, but this should be used sparingly as it is
+reduces the anonymity set, hence the recommendation that it either looks
```
```suggestion
allowed for an HTLC onion, but this should be used sparingly as it
reduces the anonymity set, hence the recommendation that it either looks
```

t-bast reviewed (commented):
Looks mostly good to me, only minor comments.
@thomash-acinq can you do a thorough review and double-check the test vectors against eclair?

thomash-acinq commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,178 @@ The _origin node_:
+  - otherwise (unknown or unset `path_id`):
+    - if the onion message is a reply to an onion message which contained a `path_id`:
+      - MUST respond (or not respond) exactly as if it did not send the initial onion message.
+  - if the `onionmsg_tlv` contains other tlv fields than `encrypted_recipient_data` and `reply_path`:
+    - MUST ignore the message.
```
Why? We actually want to use other fields, without them onion messages are not very useful.
It should be MUST ignore fields that are not expected.

rustyrussell replied:
Indeed!

I think we should ignore the message if more than one payload field is set?  If one implementation/release version accepts both invreq and inv, and another doesn't, it's a trivial way to detect?

rustyrussell replied:
```diff
diff --git a/04-onion-routing.md b/04-onion-routing.md
index c18a004..cebd8a9 100644
--- a/04-onion-routing.md
+++ b/04-onion-routing.md
@@ -1538,7 +1538,7 @@ The reader:
   - otherwise (unknown or unset `path_id`):
     - if the onion message is a reply to an onion message which contained a `path_id`:
       - MUST respond (or not respond) exactly as if it did not send the initial onion message.
-  - if the `onionmsg_tlv` contains other tlv fields than `encrypted_recipient_data` and `reply_path`:
+  - if the `onionmsg_tlv` contains more than one payload field:
     - MUST ignore the message.
   - if it wants to send a reply:
     - MUST create an onion message using `reply_path`.
```

remyers commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,178 @@ The _origin node_:
+      - MAY use `path_id` to contain a secret so it can recognize use of this `reply_path`.
+  - otherwise:
+    - MUST NOT set `reply_path`.
+- SHOULD retry via a different route if it expects a response and doesn't receive one after a reasonable period.
```
..and lines 1500 and 1501 should be combined on one line

rustyrussell commented on .aspell.en.pws:
```diff
@@ -387,9 +387,11 @@ CHECKSIGVERIFY
-unblinded
\ No newline at end of file
+unblinded
+tailer
```
No, tailer is the opposite of header.  But since you didn't know the word, I've changed it to "trailing bytes"

rustyrussell commented on bolt04/blinded-onion-message-onion-test.json:
```diff
@@ -0,0 +1,142 @@
+    "introduction_node_id": "02eec7245d6b7d2ccb30380bfbe2a3648cd7a942653f5aa340edcea1f283686619",
+    "blinding": "031195a8046dcbb8e17034bca630065e7a0982e4e36f6f7e5a8d4554e4846fcd99",
+    "hops": [
+      {
```
Fixed (and I agree with that value!)

rustyrussell commented on bolt04/blinded-onion-message-onion-test.json:
```diff
@@ -0,0 +1,142 @@
+        "privkey": "4444444444444444444444444444444444444444444444444444444444444444",
+        "onion_message": "0201025aaca62db7ce6b46386206ef9930daa32e979a35cb185a41cb951aa7d254b03c055600025550b2910294fa73bda99b9de9c851be9cbb481e23194a1743033630efba546b86e7d838d0f6e9cc0ed088dbf6889f0dceca3bfc745bd77d013a31311fa932a8bf1d28387d9ff521eabc651dee8f861fed609a68551145a451f017ec44978addeee97a423c08445531da488fd1ddc998e9cdbfcea59517b53fbf1833f0bbe6188dba6ca773a247220ec934010daca9cc185e1ce ...
+        "tlvs": {
+          "path_id": "deadbeefbadc0ffeedeadbeefbadc0ffeedeadbeefbadc0ffeedeadbeefbadc0",
```
Yeah, "encrypted_recipient_data" used to be called "path_id" in an earlier draft.  Fixed!

rustyrussell replied:
Indeed, that was from decrypting encrypted_recipient_data and looking inside, which is redundant.

rustyrussell commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,178 @@ The _origin node_:
+  - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
+  - MUST NOT include `short_channel_id`, `payment_relay` or `payment_constraints` in any `encrypted_data_tlv`
+  - MUST include `encrypted_data_tlv.next_node_id` for each non-final node.
+  - MUST NOT include any other fields in `encrypted_data_tlv` for any non-final node.
```
Yes, this is dumb: we don't need to worry about the creator of the path de-anonymizing.   Removed.

This was presumably supposed to be a requirement on onionmsg_tlv: We already say that non-final nodes should drop onionmsg_tlv which contains anything other than `encrypted_recipient_data`, but we don't say you shouldn't put anything else in onionmsg_tlv:

```diff
diff --git a/04-onion-routing.md b/04-onion-routing.md
index cebd8a9..0e95b39 100644
--- a/04-onion-routing.md
+++ b/04-onion-routing.md
@@ -1488,7 +1488,6 @@ The creator of `encrypted_recipient_data` (usually, the recipient of the onion):
   - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
   - MUST NOT include `short_channel_id`, `payment_relay` or `payment_constraints` in any `encrypted_data_tlv`
   - MUST include `encrypted_data_tlv.next_node_id` for each non-final node.
-  - MUST NOT include any other fields in `encrypted_data_tlv` for any non-final node.
   - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
 
 The writer:
@@ -1499,7 +1498,7 @@ The writer:
 - SHOULD set `onion_message_packet` `len` to 1366 or 32834.
 - SHOULD retry via a different path if it expects a response and doesn't receive one after a reasonable period.
 - For the non-final nodes' `onionmsg_tlv`:
-  - MUST NOT set `reply_path` 
+  - MUST NOT set fields other than `encrypted_recipient_data`.
 - For the final node's `onionmsg_tlv`:
   - if the final node is permitted to reply:
     - MUST set `reply_path` `blinding` to the initial blinding factor for the `first_node_id`
```

t-bast commented:
@valentinewallace can you have a look at the latest state of the PR (and ideally the test vectors)? If everything looks good, we can (finally) merge this!

ProofOfKeags commented:
I figure that this topic has been raised somewhere, but isn't the Onion Messaging system a potential DOS vector? If that discussion was resolved, is the documentation of that discussion in a different place? I am trying to catch up on the state of this and BOLT12 and IIRC this was an open point of debate for a while.

bshramin commented:
Should onion messages get added here?
https://github.com/lightning/bolts/blob/master/01-messaging.md?plain=1#L473

bshramin commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,175 @@ The _origin node_:
+as the `hop_payloads` format, except there is no "legacy" length: a 0
+`length` would mean an empty `onionmsg_payload`.
+
+Onion messages are unreliable: in particular, they are designed to
```
Since OMs are unreliable and a sender might have to retry multiple times (maybe even try numerous paths simultaneously to gain more reliability), does it make sense to add an idempotency key to onion messages?

bshramin commented on 04-onion-routing.md:
```diff
@@ -1406,6 +1407,175 @@ The _origin node_:
+- MUST set the `onion_message_packet` `version` to 0.
+- MUST construct the `onion_message_packet` `onionmsg_payloads` as detailed above using Sphinx.
+- MUST NOT use any `associated_data` in the Sphinx construcion.
+- SHOULD set `onion_message_packet` `len` to 1366 or 32834.
```
Does this mean that an OM can travel far more than 20 hops?

cdecker replied:
Potentially, yes. With varonions the limitation to fixed 20 hops went away anyway, so even the smaller 1366 byte onion could potentially contain more hops. It all deppends on the payload you want to deliver at each hop.

bshramin replied:
Thank you for the answer. So if I want to calculate the maximum number of hops I can go through, does this look right?
32834 / 66 =~ 497 hops

rustyrussell commented:
Agreed to merge pending @valentinewallace acking test vectors.

niftynei commented:
adding link to spec meeting where this was merged #1098

ProofOfKeags commented:
> I figure that this topic has been raised somewhere, but isn't the Onion Messaging system a potential DOS vector?

Responding to myself here to document the resolution of this. The current assessment is that the bandwidth of onion messages required for "proper usage" with BOLT12 is fairly low, and implementations are advised to implement some basic rate limiting to mitigate this. Rate-limiting strategy is left as an exercise to the implementer.

vincenzopalazzo commented:
>Responding to myself here to document the resolution of this. The current assessment is that the bandwidth of onion messages required for "proper usage" with BOLT12 is fairly low, and implementations are advised to implement some basic rate limiting to mitigate this. Rate-limiting strategy is left as an exercise to the implementer.

Yeah, when we have several implementation that support it we can also consider to add a line in the spec for people that will go to implement it in th future
